cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -682,26 +682,36 @@ predicate outNodeHasAddressAndIndex(
   out.getIndirectionIndex() = indirectionIndex
 }
 
-private predicate defToNode(Node nodeFrom, Def def, boolean uncertain) {
+/**
+ * INTERNAL: Do not use.
+ *
+ * Holds if `node` is the node that corresponds to the definition of `def`.
+ */
+predicate defToNode(Node node, Def def, SourceVariable sv, IRBlock bb, int i, boolean uncertain) {
+  def.hasIndexInBlock(bb, i, sv) and
   (
-    nodeHasOperand(nodeFrom, def.getValue().asOperand(), def.getIndirectionIndex())
+    nodeHasOperand(node, def.getValue().asOperand(), def.getIndirectionIndex())
+    or
+    nodeHasInstruction(node, def.getValue().asInstruction(), def.getIndirectionIndex())
     or
-    nodeHasInstruction(nodeFrom, def.getValue().asInstruction(), def.getIndirectionIndex())
+    node.(InitialGlobalValue).getGlobalDef() = def
   ) and
   if def.isCertain() then uncertain = false else uncertain = true
 }
 
 /**
  * INTERNAL: Do not use.
  *
- * Holds if `nodeFrom` is the node that correspond to the definition or use `defOrUse`.
+ * Holds if `node` is the node that corresponds to the definition or use at
+ * index `i` in block `bb` of `sv`.
+ *
+ * `uncertain` is `true` if this is an uncertain definition.
  */
-predicate nodeToDefOrUse(Node nodeFrom, SsaDefOrUse defOrUse, boolean uncertain) {
-  // Node -> Def
-  defToNode(nodeFrom, defOrUse, uncertain)
+predicate nodeToDefOrUse(Node node, SourceVariable sv, IRBlock bb, int i, boolean uncertain) {
```

Comment:
Here and in other places: what's the intuition behind adding `sv`, `bb`, and `i` as arguments?

Reply:
A use (or a def) is basically defined by the triple `(bb, i, sv)` (i.e., a basic block, an index and a variable). The variable is necessary because a specific control-flow position (i.e., the `(bb, i)` part) can define (or use) multiple variables. For example, if you have:
```cpp
p = new int(0);
```
that's both a definition of `p` (i.e., some address is written to `p`) and a definition of `*p` (i.e., the value `0` is written to `*p`).

Similarly, for a function call such as `use(p)` there's a use of `p` and `*p` at the same index in the same basic block.

Previously, we materialized the triple `(bb, i, sv)` in a newtype called `SsaDefOrUse` that contained:
- All uses, and
- All defines that could reach a use

However, that meant that we had two different-but-related classes for definitions: one called `SsaDefOrUse`, and one called `Definition`. The former is the newtype we created and the latter is the newtype created by the shared SSA library.

By getting rid of the `SsaDefOrUse` class we're left with just one `Definition` class (i.e., the one produced by the shared SSA library), and we don't need to convert back and forth between the two.

Reply:
Thanks. That helps in understanding what's going on here.